### PR TITLE
Parser

### PR DESCRIPTION
**Pull request by bidzyyys**

### CODE REVIEW

File: entity_linking/grammar/analysis.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        self._add_info1 = add_info_1
+        self._add_info2 = add_info_2
+
+    def __eq__(self, other):
+        return self.get_lemma() == other.get_lemma() and \
+            self.get_tag() == other.get_tag() and \
+            self.get_add_info_1() == other.get_add_info_1() and \
+            self.get_add_info_2() == other.get_add_info_2()
+
+    def __ne__(self, other):
```
Comment (pik694) [inline entity_linking/grammar/analysis.py]: I believe that this is the default implementation, therefore you do not have to reimplement it

File: entity_linking/grammar/token.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        self._word = analysis[0]
+        self._analyses = list()
+        self.add_analysis(analysis=analysis)
+
+    def __eq__(self, other):
+        return self.get_range() == other.get_range() and \
+            self.get_word() == other.get_word() and \
+            self.get_analyses() == other.get_analyses()
+
+    def __ne__(self, other):
```
Comment (pik694) [inline entity_linking/grammar/token.py]: as above

File: entity_linking/grammar/range.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+class Range():
+    def __init__(self, start: int, end: int):
+        self._start = start
+        self._end = end
+
+    def __eq__(self, other):
+        return self.get_start() == other.get_start() and \
+            self.get_end() == other.get_end()
+
+    def __ne__(self, other):
```
Comment (pik694) [inline entity_linking/grammar/range.py]: same

File: entity_linking/grammar/analysis.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        return not self == other
+
+    def get_lemma(self) -> str:
```
Comment (pik694) [inline entity_linking/grammar/analysis.py]: all such getters can be refactored to become properties

File: tests/grammar/test_analysis.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
+from pytest import raises
+from entity_linking.grammar.analysis import Analysis, AnalysisError
+
+
+def test_eq_other_lemma():
```
Comment (pik694) [inline tests/grammar/test_analysis.py]: what about parametrising one test - less code, less duplication, less tests that test just one function (`eq`)

File: tests/grammar/test_analysis.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    tag = "c"
+    add_info_1 = ["a"]
+    add_info_2 = ["b"]
+    analysis1 = Analysis(lemma=lemma1, tag=tag,
+                         add_info_1=add_info_1, add_info_2=add_info_2)
+    analysis2 = Analysis(lemma=lemma2, tag=tag,
+                         add_info_1=add_info_1, add_info_2=add_info_2)
+
+    assert not analysis1 == analysis2
+    assert analysis1 != analysis2
```
Comment (pik694) [inline tests/grammar/test_analysis.py]: this check is obsolete if you use default implementation of `__neq__`

File: tests/grammar/test_analysis.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+def test_from_analysis():
+    word = "a"
+    lemma = "a"
+    tag = "b"
+    add_info_1 = ["a"]
+    add_info_2 = ["2"]
+    analysis_tuple = (word, lemma, tag, add_info_1, add_info_2)
+    analysis = Analysis.from_analysis(analysis=analysis_tuple)
+    assert analysis.get_lemma() == lemma
```
Comment (pik694) [inline tests/grammar/test_analysis.py]: I would rather create an `Analysis` object using constructor and then just compare the two instances

File: tests/grammar/test_range.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    assert not r1 == r2
+    assert r1 != r2
+
+def test_equal():
+    start = 1
+    end = 3
+    r1 = Range(start=start, end=end)
+    r2 = Range(start=start, end=end)
+    assert r1 == r2
+    assert not r1 != r2
```
Comment (pik694) [inline tests/grammar/test_range.py]: comments from the previous file apply here too

File: tests/grammar/test_token.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+
+def test_eq_token(token):
+    token1 = Token.from_analysis(
+        start=token.get_range().get_start(),
+        end=token.get_range().get_end(),
+        analysis=(token.get_word(), token.get_analyses()[0].get_lemma(), token.get_analyses()[0].get_tag(), token.get_analyses()[0].get_add_info_1(), token.get_analyses()[0].get_add_info_2())
+    )
+    assert token == token1
+    assert not token != token1
```
Comment (pik694) [inline tests/grammar/test_token.py]: please remove all such checks

File: tests/parser/test_parser.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+def double_analysis_token():
+    token = Token.from_analysis(start=0, end=1, analysis=(
+        'Zaginieni', 'zaginiony:a', 'adj:pl:nom.voc:m1:pos', [], []))
+    token.add_analysis(('Zaginieni', 'zaginiony:s',
+                        'subst:pl:nom.voc:m1', ['nazwa_pospolita'], []))
+    return token
+
+
+def test_analyse_text(parser):
+    word = "tests"
```
Comment (pik694) [inline tests/parser/test_parser.py]: morfeusz is capable of parsing polish text ...

File: tests/parser/test_parser.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    assert parser.analyse_text(word) == Morfeusz().analyse(word)
```
Comment (pik694) [inline tests/parser/test_parser.py]: this is basically what `analyse_text`  method does underneath - we do not want to test such things

File: entity_linking/grammar/analysis.py (hunk 1 above)
Comment (bidzyyys) [reply entity_linking/grammar/analysis.py]: After removal, assertion fails.

File: entity_linking/grammar/token.py (hunk 2 above)
Comment (bidzyyys) [reply entity_linking/grammar/token.py]: As above

File: entity_linking/grammar/range.py (hunk 3 above)
Comment (bidzyyys) [reply entity_linking/grammar/range.py]: same

**Pull request merged**